compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -3060,7 +3068,39 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 (root, pred) if seen_preds.contains(&pred) || seen_preds.contains(&root) => {}
                 _ => continue,
             }
-            error.obligation.cause.span = span;
+            error.obligation.cause.span = idx_span;
+
+            // If the index value is a double borrow, that can cause typeck errors
+            // that can be easily resolved by removing the borrow from the expression.
+            // We check for that here and provide a suggestion in a custom obligation
+            // cause code.
+            if let hir::ExprKind::AddrOf(_, _, idx) = idx.kind {
+                let idx_t = self.typeck_results.borrow().expr_ty(idx);
+                if let Some((index_ty, _element_ty)) =
+                    self.lookup_indexing(expr, base, base_t, idx, idx_t)
+                {
+                    let (_ty, err) =
+                        self.demand_coerce_diag(idx, idx_t, index_ty, None, AllowTwoPhase::No);
```

Comment:
I think you should manually recreate the logic in `lookup_indexing`, and if you need to process obligations, you should throw them into your *own* `ObligationCtxt`.

As for the `demand_coerce` call, I think it's probably best for us to approximate it with a `ObligationCtxt::eq` call instead, to again avoid poisoning our own fulfillment context.

Essentially what I want to avoid is that `fulfillment_ctxt.select_where_possible`. It's not great, and unless you can convince me that there's a good reason that we can't approximate it with less invasive machinery, then I think we should try our best to avoid it.

Reply:
What do you think of creating a new fulfillment context for this check instead?

Reply:
either works, though I think `ObligationCtxt` has an easier to use API